### DISCUSSION TITLE
How to subscribe to IsEffectivelyVisible changed

### DISCUSSION THREAD

**Discussion by danipen** [AvaloniaUI/Avalonia#7876] [Ideas]
I'm writing a rotator component that updates some properties every frame. Obviously, I don't want to update properties when the component is not visible to avoid wasting CPU.

# Issue

I'm trying to figure out how to manage if a `Control` changed its visibility, and it's being a little bit tricky because this component is not being hidden itself directly (it's part of the ControlTemplate). **So the component is indirectly hidden when some of its parents is hidden.**

# Workaroud

I noticed that there is an `IsEffectivelyVisible` property that is useful to determine the visibility in this scenario but was unable to be notified when the EffectivelyVisibility was changed.

I made a sample application to test indirect visibility, and I noticed that when a control is indirectly set to visible/hidden, the property `TransformedBounds` is changed, so I wrote this ugly workaround:

```csharp
protected override void OnPropertyChanged<T>(AvaloniaPropertyChangedEventArgs<T> change)
{
    base.OnPropertyChanged(change);

    if (change.Property == TransformedBoundsProperty || change.Property == IsVisibleProperty)
    {
        if (IsEffectivelyVisible)
            Start(this);
       else
            Stop(this);
    }
}
```

# Questions
So I'm missing an `IsEffectivelyVisibleProperty` that would really be useful for this scenario. 

- How difficult would be to implement IsEffectivelyVisibleProperty?
- Is any other (better) way to do this?
- Any ideas / comments?

**Comment by danipen** [comment]
There is an interesting discussion about something similar for WPF, here: https://github.com/microsoft/microsoft-ui-xaml/issues/674

**Reply by robloo** [reply]
That discussion was actually for UWP/WinUI instead of WPF but it's the practically the same problem encountered in all XAML frameworks to date.

**Reply by danipen** [reply]
Ohh sorry... yes you're right I misread WPF...

So... No easy solution?

**Reply by robloo** [reply]
As far as I know, you are correct, there is no easy solution. Either you redesign your app so the situation never happens or you have some specialized code like you have.

I think adding a new IsUserVisible property would be great to handle all this. IsEffectivelyVisible won't tell you if an item is scrolled off screen for example.
